at-angular.ts

```diff
@@ -107,8 +151,36 @@ module at {
             if (target.$inject && target.$inject.length > 0) {
                 factory.$inject = target.$inject.slice(0);
             }
-            angular.module(moduleName).factory(className, factory);
+            angular.module(moduleName).factory(className || getFuncName(target), factory);
+        };
+    }
+
+    export function decorator(moduleName: string, targetProvider: string): at.IClassAnnotationDecorator {
+
+        return (targetClass: any): void => {
+
+            angular
+                .module(moduleName)
+                .config([
+                    '$provide',
+                    function($provide: angular.auto.IProvideService): void {
+
+                    $provide.decorator(targetProvider, [
+                       '$delegate',
+                        ($delegate: any): void => {
+
+                            $delegate[targetProvider] = $delegate;
+
+                            angular.extend($delegate, new targetClass());
```

Comment:
Ok, sure there is need of injection. This complicates things a lot, but this is possible :)

If there are any `@inject(…)`  annotations for constructor args or for whole class, there will also exist static member `$inject` on a class :) So then you can:

```typescript
angular.module(moduleName).config(['$provide', ($provide: angular.auto.IProvideService): void => {

    delegation.$inject = targetClass.$inject.map((inject: string) => inject === '$delegate' ? targetProvider : inject);
    function delegation(...injects: string[]) {
        return angular.extend($delegate, targetClass.prototype, new targetClass(...injects));
    }

    $provide.decorator(targetProvider, delegation);

}]);
```

If this does not work - than we should probably use something like `$provide.instantiate` instead of `new targetClass(...injects)`. I'm not sure.